### PR TITLE
ARM exclusive monitor and multicore fixes

### LINKED ISSUE #396: ARM Exclusive Monitor is not being cleared when it should
Any store to a "exclusive" region should clear the exclusive monitor, however this is not happening currently, the only stores that are clearing the exclusive monitor are the exclusive instructions.

The main problem is that the clear exclusive monitor + store operation should be atomic, and I couldn't think yet in a way to do that without slowing down all stores (due to the use of a lock being necessary).

### PR DESCRIPTION
**Pull request by gdkchan**
This implements the arm exclusive load/store instruction using the cmpxchg instruction on x64. This is more efficient and less error prone than the previous approach using locks, it doesn't exactly match what an ARM CPU does, but should be good enough. Unfortunately, theres no intrinsic for the cmpxchg16b instruction (necessary for 128-bits compare exchange) on .NET, so I had to include the raw code for a function with this instruction that does the operation, and then call it. This requires us to implement platform specific code to map the memory as executable.

Additionally, also enable multicore scheduling by default, as the previous problems that it caused (deadlocks and other errors on some games) seems to be fixed.

~~Still needs Linux/Mac support, so this is still WIP.~~

Fixes #396

### CODE REVIEW

**Comment by gdkchan** [comment]
Ready for review.

**Comment by gdkchan** [comment]
I added the cmpxchg16b cpuid flag check, now it needs to be re-tested on linux.

**Comment by jduncanator** [comment]
This seems like a perfect candidate for some unit tests, ensuring that the logic is correct in most instances. You could probably write some small unit tests based off of code seen in games, but also would be nice to write some tests around the logic defined in the ARM specification.

Either way, I'm happy for that to be a job for another day...

File: ChocolArm64/Memory/MemoryManager.cs [hunk 1]
```diff
@@ -75,104 +58,139 @@ public MemoryManager(IntPtr ram)
...
-            {
-                Monitor.Exit(_monitors);
+            int expected32 = (currentValue & ~byte.MaxValue) | expected;
+            int desired32  = (currentValue & ~byte.MaxValue) | desired;
 
-                return false;
+            return Interlocked.CompareExchange(ref *ptr, desired32, expected32) == expected32;
+        }
+
+        public bool AtomicCompareExchangeInt16(long position, short expected, short desired)
```
[the remark this answers is not in the document]
Comment (gdkchan) [reply ChocolArm64/Memory/MemoryManager.cs]: On all cases I've seen, the store exclusive instruction are used in a loop, where it tries again in the case of failure. I was not able to think in any other way to do CAS of 16 or 8-bits values, as x86 only provides instructions for 32-bits, 64-bits and 128-bits CAS.

[the remark this answers is not in the document]
Comment (gdkchan) [reply ChocolArm64/Memory/MemoryManager.cs]: Correction -- 8 and 16-bits CAS is supported on x86, but not exposed through .NET `Interlocked.CompareExchange` methods, so we'd need even more native code to use them. I can look into implementing it if desired.

[the remark this answers is not in the document]
Comment (jduncanator) [reply ChocolArm64/Memory/MemoryManager.cs]: No worries. Probably not worth the additional native code at this point in time. If we determine that it's a cause of issue, we can always add native implementations for the other CAS instructions as needed.

**Comment by jduncanator** [comment]
> I added the cmpxchg16b cpuid flag check, now it needs to be re-tested on linux.

CPUID implementation checks out on Linux and Windows, so should be good there.

**Pull request merged**